### ISSUE TITLE
float value bug when quoting strings only-when-needed

### ISSUE THREAD

**Issue by andreineculau** [adrienverge/yamllint#503]
The config

```yml
extends: default
rules:
  float-values:
    forbid-inf: true
    forbid-nan: true
    forbid-scientific-notation: true
    require-numeral-before-decimal: true
  quoted-strings:
    quote-type: single
    required: only-when-needed
    allow-quoted-quotes: true
```

will invalidate

```yml
color: '5319E7'
color: 5319E7
```

with

```text
./test.yml:1:8: [error] string value is redundantly quoted with single quotes (quoted-strings)
./test.yml:2:8: [error] forbidden scientific notation "5319E7" (float-values)
```

**Comment by adrienverge** [comment]
Hello, thanks for the report!

I guess the bug comes from PyYAML (the YAML parser behind yamllint) not converting `5319E7` (nor `5319e7`) to a float, while according to [the YAML specification](https://yaml.org/spec/1.2.2/#10214-floating-point) it should:
```
-? ( 0 | [1-9] [0-9]* ) ( \. [0-9]* )? ( [eE] [-+]? [0-9]+ )?
```

Hence, yamllint creates a false positive on "*string value is redundantly quoted with single quotes*", because it sees both `5319E7` and `'5319E7'` having the same "string" type.

I see 2 ways to fix this:
- Fix PyYAML.
- Add a workaround in yamllint to fix the type inference, similarly to commit 1b378ed "quoted-strings: Fix explicit octal recognition". See pull request https://github.com/adrienverge/yamllint/pull/354 that fixed issue https://github.com/adrienverge/yamllint/issues/351.

@DerekTBrown do you want to fix it?

**Comment by perlpunk** [comment]
yamllint/pyyaml are correct here. They implement YAML 1.1.

The YAML 1.1 spec requires at least a dot: https://yaml.org/type/float.html

For better comparison have a look here:
* [YAML Schemas/Types per version](https://perlpunk.github.io/yaml-test-schema/schemas.html)
* [Example data](https://perlpunk.github.io/yaml-test-schema/data.html)

The example data is actually already used in pyyaml:
* https://github.com/yaml/pyyaml/blob/master/tests/lib/test_schema.py
* https://github.com/yaml/pyyaml/blob/master/tests/data/yaml11.schema

And it passes all float tests.

**Comment by adrienverge** [comment]
@perlpunk thanks! It's clearer now why PyYAML doesn't convert `5319e7` to a float.

My point of view:
- There remains a problem with yamllint, because @andreineculau's first code snippet makes yamllint complain both *with* or *without* quotes.
- Yamllint would "prefer" parsing the latest YAML version (1.2.2) if easily possible. It currently parses YAML 1.1 only because PyYAML does so. Thus, I would be favorable to a fix in yamllint to "patch" parsing to behave like 1.2.
  That's what we've already done for issue https://github.com/adrienverge/yamllint/issues/351.
